### PR TITLE
Enable multiple poll-mode-drivers for DPDK testing

### PR DESCRIPTION
**Pull request by mcgov**
DPDK testing in v2 checks multiple poll mode driver configurations, this PR adds the ability to run previous tests (verify build and verify send-receive) with the default failsafe PMD and by using the netvsc pmd directly.

### CODE REVIEW

File: lisa/nic.py [hunk 1]
```diff
@@ -165,6 +174,21 @@ def _get_nic_device(self, nic_name: str) -> str:
...
 
+    def _get_nic_uuid(self, nic_name: str) -> str:
+        full_dev_path = self._node.execute(
+            f"readlink /sys/class/net/{nic_name}/device",
+            expected_exit_code_failure_message=(
+                f"could not get sysfs device info for {nic_name}"
+            ),
+        )
+        uuid = os.path.basename(full_dev_path.stdout.strip())
+        self._node.log.info(f"{nic_name} UUID:{uuid}")
```
Comment (squirrelsc) [inline lisa/nic.py]: Avoid info level log in the underlying package, unless it tells a story in the info level log.

File: lisa/nic.py [hunk 2]
```diff
@@ -247,3 +271,56 @@ def _get_host_if_info(self) -> None:
...
+            nic.dev_uuid, f"/sys/bus/vmbus/drivers/{driver_module}/unbind", sudo=True
+        )
+
+    def _bind(self, nic: NicInfo, driver_module: str) -> None:
+        echo = self._node.tools[Echo]
+        echo.write_to_file(
+            nic.dev_uuid, f"/sys/bus/vmbus/drivers/{driver_module}/bind", sudo=True
+        )
+
+    def bind_to_pmd(self, nic: NicInfo, pmd: str) -> None:
```
Comment (squirrelsc) [inline lisa/nic.py]: Is this related to dpdk? If so, it should be in test case or dpdk tools.

Comment (mcgov) [reply lisa/nic.py]: The PMD nomenclature is specific to the dpdk tools, I can move this part into the test suite itself.

Comment (squirrelsc) [reply lisa/nic.py]: the `_bind/_unbind` can be public, if they need to be accessed outside.

File: lisa/nic.py [hunk 3]
```diff
@@ -247,3 +271,56 @@ def _get_host_if_info(self) -> None:
...
+
+    def _unbind(self, nic: NicInfo, driver_module: str) -> None:
+        echo = self._node.tools[Echo]
+        echo.write_to_file(
+            nic.dev_uuid, f"/sys/bus/vmbus/drivers/{driver_module}/unbind", sudo=True
+        )
+
+    def _bind(self, nic: NicInfo, driver_module: str) -> None:
+        echo = self._node.tools[Echo]
+        echo.write_to_file(
```
Comment (squirrelsc) [inline lisa/nic.py]: If it needs to bind/unbind multiple nics, can this method support?

File: lisa/nic.py [hunk 4]
```diff
@@ -247,3 +271,56 @@ def _get_host_if_info(self) -> None:
                 self.get_nic(nic).mac_addr = mac_addr
             else:
                 fail(f"Could not parse output of ip addr show {nic}")
+
+    def enable_uio_hv_generic_for_nic(self, nic: NicInfo) -> None:
+        # using netvsc pmd directly for dpdk on hv counterintuitively requires
```
Comment (squirrelsc) [inline lisa/nic.py]: move to dpdk.

File: lisa/nic.py [hunk 5]
```diff
@@ -89,6 +93,10 @@ class Nics(InitializableMixin):
         r"([a-zA-Z0-9]+)"  # capture device
     )
 
+    # hv_uio_generic driver uuid, a constant value used by vmbus.
+    # https://doc.dpdk.org/guides/nics/netvsc.html#installation
+    HV_UIO_GENERIC_UUID = "f8615163-df3e-46c5-913f-f2d2f965ed0e"
```
Comment (squirrelsc) [inline lisa/nic.py]: move to dpdk.

File: microsoft/testsuites/dpdk/dpdktestpmd.py [hunk 6]
```diff
@@ -197,20 +197,33 @@ def _install_dependencies(self) -> bool:
...
             )
         )
-        return (
-            f'--vdev="net_vdev_netvsc{vdev_id},iface={node_nic.upper}"'
-            f' --allow "{node_nic.pci_slot}"'
-        )
+        vdev_info = ""
+        if pmd == "failsafe":
+            vdev_info = f'--vdev="net_vdev_netvsc{vdev_id},iface={node_nic.upper}"'
+        elif pmd == "netvsc":
```
Comment (squirrelsc) [inline microsoft/testsuites/dpdk/dpdktestpmd.py]: How about else?

File: lisa/nic.py (hunk 3 above)
Comment (mcgov) [reply lisa/nic.py]: It binds a single one at a time but we could implement a bind_multiple version if we ever need it. I don't need it right now though.

File: lisa/nic.py (hunk 1 above)
Comment (mcgov) [reply lisa/nic.py]: It's nice to have the data labeled with what it is, I'll move it to debug level

File: lisa/nic.py (hunk 3 above)
Comment (squirrelsc) [reply lisa/nic.py]: It overwrites the file every time. Even the nics are enabled one by one, the last one will overwrite previous ones.

Comment (mcgov) [reply lisa/nic.py]: I don't think it works like that. The 'bind' file is not a real file, it's just created for communication with the driver, so writing a new device ID into the file shouldn't unbind the other devices bound to the driver.

Comment (squirrelsc) [reply lisa/nic.py]: Thank you for explaining, it's ok to me now.

File: lisa/nic.py [hunk 7]
```diff
@@ -247,3 +267,15 @@ def _get_host_if_info(self) -> None:
                 self.get_nic(nic).mac_addr = mac_addr
             else:
                 fail(f"Could not parse output of ip addr show {nic}")
+
+    def unbind(self, nic: NicInfo, driver_module: str) -> None:
```
Comment (squirrelsc) [inline lisa/nic.py]: move public methods above the private methods.

File: microsoft/testsuites/dpdk/dpdktestpmd.py [hunk 8]
```diff
@@ -197,20 +197,40 @@ def _install_dependencies(self) -> bool:
...
         )
-        return (
-            f'--vdev="net_vdev_netvsc{vdev_id},iface={node_nic.upper}"'
-            f' --allow "{node_nic.pci_slot}"'
-        )
+        vdev_info = ""
+        if node_nic.bound_driver == "hv_netvsc":
+            vdev_info = f'--vdev="net_vdev_netvsc{vdev_id},iface={node_nic.upper}"'
+        elif node_nic.bound_driver == "uio_hv_generic":
+            pass
```
Comment (squirrelsc) [inline microsoft/testsuites/dpdk/dpdktestpmd.py:216]: It looks the `...` is the new fashion.

Comment (mcgov) [reply microsoft/testsuites/dpdk/dpdktestpmd.py:216]: I think `pass` is more readable. It's a well-known keyword and it shows up highlighted since it's a reserved keyword.

**Review by squirrelsc — APPROVED** [review]
